### PR TITLE
Make the sum fold numerically stable

### PR DESCRIPTION
**Pull request by adithyaov**

### CODE REVIEW

File: core/src/Streamly/Internal/Data/Fold.hs [hunk 1]
```diff
@@ -310,6 +310,7 @@ import qualified Data.Set as Set
 import qualified Prelude
 import qualified Streamly.Internal.Data.Array.Foreign.Mut.Type as MA
 import qualified Streamly.Internal.Data.Array.Foreign.Type as Array
+import qualified Streamly.Internal.Data.Fold.Window as WFold
```
Comment (harendra-kumar) [inline core/src/Streamly/Internal/Data/Fold.hs:313]: FoldW - starting with prefix Fold is better.

File: core/src/Streamly/Internal/Data/Fold.hs [hunk 2]
```diff
@@ -683,7 +684,7 @@ countDistinctInt = fmap (\(Tuple' _ n) -> n) $ foldl' step initial
 -- @since 0.7.0
 {-# INLINE sum #-}
 sum :: (Monad m, Num a) => Fold m a a
-sum =  foldl' (+) 0
+sum = lmap (\a -> (a, Nothing)) WFold.sum
```
Comment (harendra-kumar) [inline core/src/Streamly/Internal/Data/Fold.hs:687]: `sum = FoldW.cumulative FoldW.sum`

Comment (harendra-kumar) [reply core/src/Streamly/Internal/Data/Fold.hs:687]: Change the documentation.

```
>>> sum = FoldW.cumulative FoldW.sum
```

Same as following but numerically stable:

```
>>> sum =  Fold.foldl' (+) 0
>>> sum = fmap getSum $ Fold.foldMap Sum
```

**Review by harendra-kumar — CHANGES_REQUESTED** [review]

**Comment by harendra-kumar** [comment]
Need to fix benchmark regression before merging this.

**Comment by harendra-kumar** [comment]
Fix the doctests.

File: core/src/Streamly/Internal/Data/Fold.hs [hunk 3]
```diff
@@ -683,12 +685,17 @@ countDistinctInt = fmap (\(Tuple' _ n) -> n) $ foldl' step initial
 -- identity (@0@) when the stream is empty. Note that this is not numerically
 -- stable for floating point numbers.
 --
--- > sum = fmap getSum $ Fold.foldMap Sum
+-- >>> sum = FoldW.cumulative FoldW.sum
+--
+-- Same as following but numerically stable:
+--
+-- >>> sum = Fold.foldl' (+) 0
+-- >>> sum = fmap Data.Monoid.getSum $ Fold.foldMap Data.Monoid.Sum
```
Comment (harendra-kumar) [inline core/src/Streamly/Internal/Data/Fold.hs:693]: You could just say "import Data.Monoid (getSum, Sum)" on the previous line, the fully qualified names can make the snippet less readable.

**Review by harendra-kumar — APPROVED** [review]

File: core/src/Streamly/Internal/Data/Fold.hs (hunk 3 above)
Comment (adithyaov) [reply core/src/Streamly/Internal/Data/Fold.hs:693]: They were fully qualified imports in other parts of the code. So I assumed it was like that for a reason.

Comment (adithyaov) [reply core/src/Streamly/Internal/Data/Fold.hs:693]: I guess this is fine. The imported module isn't too long.

**Pull request merged**